Adding LLM based Masking Mutation Operator, and Combiend search

Andydiantu opened the pull request.

AnnieFraz commented on src/main/java/gin/Patch.java:
```diff
@@ -180,7 +180,18 @@ public void addRandomEditOfClass(Random rng, Class<? extends Edit> allowableEdit
+     * @param allowableEditTypes - list of allowable edit types
+     * @param combined_probability - probability of choosing a combined edit
+     */
+    public void addRandomEditOfClasses(Random rng, List<Class<? extends Edit>> allowableEditTypes, double combined_probability) {
```
For code quality, why are we having snake case and camelCase?

AnnieFraz commented on src/main/java/gin/Patch.java:
```diff
@@ -212,11 +223,25 @@ private Edit randomEdit(Random rng, List<Class<? extends Edit>> allowableEditTyp
+        Logger.info("Choosing an edit type from: " + allowableEditTypes);
         Class<? extends Edit> editType = allowableEditTypes.get(rng.nextInt(allowableEditTypes.size()));
 
+        if(allowableEditTypes.size() == 5){
```
Please add a comment of why 5 here for future people who are working on the project?

AnnieFraz commented on src/main/java/gin/edit/llm/LLMMaskedStatement.java:
```diff
@@ -0,0 +1,274 @@
+        Statement statementToMask = drawStatementFromSourceFile(sf, (rng != null ? rng : new Random()));
+        Logger.info( "Statement to mask: " + statementToMask.toString());
+
+        LLMQuery llmQuery;
```
do we want this to be something in case it does not fit the below if statment?

Andydiantu replied:
Hi, this is just selecting the one statement that will be masked from target method

sandybrownlee replied:
LLM query will be populated as there's an else block to pick up any other cases.

AnnieFraz commented on src/main/java/gin/edit/llm/LLMMaskedStatement.java:
```diff
@@ -0,0 +1,274 @@
+
+
+    public Statement drawStatementFromSourceFile(SourceFileTree sourceFileTree, Random rng){
+        //TODO choose meaningfull statemetn to mask, 
```
```suggestion
        //TODO choose meaningfull statement to mask, 
```

AnnieFraz commented on src/main/java/gin/edit/llm/LLMMaskedStatement.java:
```diff
@@ -0,0 +1,274 @@
+        return this.getClass().getCanonicalName() + " \"" + destinationFilename + "\":" + destinationStatement + "\nPrompt: !!!\n" + lastPrompt +  "\n!!! --> !!!\n" + lastReplacement + "\n!!!";
+    }
+
+
```
whitespace

AnnieFraz commented on src/main/java/gin/util/GP.java:
```diff
@@ -50,6 +50,9 @@ public abstract class GP extends Sampler {
     @Argument(alias = "is", description = "Random seed for individual selection")
     protected Integer individualSeed = 123;
 
+    @Argument(alias = "pb", description = "Probablity of combiend")
```
```suggestion
    @Argument(alias = "pb", description = "Probability of combined")
```

Andydiantu commented:
Hi, I have updated the pull request based on your comment. 

I also updated the code including moving the patch mutation type selection from Patch Class to two search algorithms, so now the patch class will have no differences to the original LLM branch. 

Could you review it again? 
Thank you
